### PR TITLE
Add comprehensive input validation with detailed error responses

### PR DESCRIPTION
**Pull request by olaservo**
## Summary
- Added comprehensive input validation middleware for all Todo endpoints
- Implemented field-specific error messages with structured error responses  
- Added validation for title length, HTML tag prevention, and type checking
- Enhanced ID parameter validation for route security
- Added comprehensive test coverage for all validation scenarios

### New Files
- `src/middleware/validation.ts` - Core validation logic with custom error handling
- `src/middleware/validation.test.ts` - Comprehensive test suite (15+ test cases)

### Modified Files
- `src/routes/todos.ts` - Integrated validation middleware, improved error responses
- `README.md` - Added validation rules and error response format documentation
- `package.json` - Added supertest for API testing

### Validation Rules Implemented
- **Title**: Required, 1-100 chars, no HTML tags, string type validation
- **Description**: Optional, max 500 chars, no HTML tags, string type validation  
- **Completed**: Boolean type validation when provided
- **ID Parameters**: Positive integer validation for all routes

## Test Coverage
- Valid input acceptance
- Missing required fields
- Invalid data types
- Length limit enforcement
- HTML tag prevention (XSS protection)
- Multiple simultaneous errors
- Edge cases and boundary conditions

## API Response Changes
**Before**: Simple error strings
```json
{ "error": "Title is required" }
```

**After**: Structured error objects with field details
```json
{
  "error": "Validation failed", 
  "details": [
    { "field": "title", "message": "Title is required" },
    { "field": "description", "message": "Description must be 500 characters or less" }
  ]
}
```

## Security Improvements
- HTML tag prevention to mitigate XSS attacks
- Input sanitization with automatic trimming
- Type validation to prevent injection attacks
- ID format validation to prevent path traversal

## Testing Instructions

1. **Install dependencies**: `npm install`
2. **Run test suite**: `npm test` (should show 15+ passing tests)
3. **Test API manually**:
   ```bash
   npm run dev
   
   # Test valid creation
   curl -X POST localhost:3000/api/todos \
     -H "Content-Type: application/json" \
     -d '{"title":"Valid Todo","description":"Valid description"}'
   
   # Test validation errors
   curl -X POST localhost:3000/api/todos \
     -H "Content-Type: application/json" \
     -d '{"title":"","description":"'$(printf 'a%.0s' {1..501})'"}'
   ```

## Breaking Changes
**Error response format has changed** - API consumers will need to handle the new structured error format

## Future Enhancements
- [ ] Add rate limiting validation
- [ ] Implement custom validation rules configuration
- [ ] Add internationalization for error messages
- [ ] Consider adding request sanitization middleware

### CODE REVIEW

File: src/middleware/validation.ts [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+    errors.push({ field: 'title', message: 'Title cannot contain HTML tags' });
+  }
+
+  // Validate description
+  if (description !== undefined && description !== null) {
+    if (typeof description !== 'string') {
+      errors.push({ field: 'description', message: 'Description must be a string', value: description });
+    } else if (description.length > 500) {
+      errors.push({ field: 'description', message: 'Description must be 500 characters or less' });
+    } else if (/<.*>/g.test(description)) {  // BUG: Inconsistent regex - allows <img> but title rejects it
```
Comment (dandeliongold) [inline src/middleware/validation.ts:42]: **Bug: Inconsistent HTML validation regex**

Line 42 uses `/<.*>/g` while line 32 uses `/<[^>]*>/g`. This inconsistency means:
- Title rejects: `` ✅
- Description allows: `` ❌ (should be rejected)

**Fix:** Use the same regex pattern for both:
```typescript
} else if (/<[^>]*>/g.test(description)) {
```

File: src/middleware/validation.ts [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+      details: errors
+    });
+  }
+
+  next();
+};
+
+export const validateTodoId = (req: Request, res: Response, next: NextFunction) => {
+  const { id } = req.params;
+
```
Comment (dandeliongold) [inline src/middleware/validation.ts:102]: **Bug: Missing validation for ID "0"**

The regex `/^\d+$/` allows "0" as a valid ID, which may not be intended if your ID generation starts from 1.

**Consider:** Adding explicit check for positive integers:
```typescript
const numericId = parseInt(id, 10);
if (!id || !/^\d+$/.test(id) || numericId <= 0) {
```

File: src/routes/todos.ts [hunk 3]
```diff
@@ -9,54 +10,67 @@ router.get('/', (req: Request, res: Response) => {
...
+router.post('/', validateTodoCreate, (req: Request, res: Response) => {
   const { title, description }: CreateTodoRequest = req.body;
   
-  if (!title || title.trim() === '') {
-    return res.status(400).json({ error: 'Title is required' });
-  }
-  
-  const todo = todoStore.create({ title: title.trim(), description });
+  // BUG: No try-catch block - if todoStore.create throws, server will crash
+  const todo = todoStore.create({
```
Comment (dandeliongold) [inline src/routes/todos.ts:31]: **Bug: Potential server crash**

If `todoStore.create()` throws an error, it will crash the server since there's no error handling.

**Fix:** Add try-catch block:
```typescript
try {
  const todo = todoStore.create({ 
    title: title.trim(), 
    description: description?.trim() || undefined 
  });
  res.status(201).json(todo);
} catch (error) {
  res.status(500).json({ error: 'Failed to create todo' });
}
```

File: src/middleware/validation.test.ts [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+        .post('/')
+        .send({ title: '<script>alert("xss")</script>' });
+
+      expect(response.status).toBe(400);
+      expect(response.body.details).toContainEqual({
+        field: 'title',
+        message: 'Title cannot contain HTML tags'
+      });
+    });
+
```
Comment (dandeliongold) [inline src/middleware/validation.test.ts:61]: **Missing test coverage**

The test comment indicates missing test for description HTML validation. This should be added to ensure both fields have consistent validation behavior.

**Suggested test:**
```typescript
it('should reject description with HTML tags', async () => {
  const response = await request(app)
    .post('/')
    .send({ title: 'Valid Title', description: '<script>alert("xss")</script>' });

  expect(response.status).toBe(400);
  expect(response.body.details).toContainEqual({
    field: 'description',
    message: 'Description cannot contain HTML tags'
  });
});
```

File: src/middleware/validation.test.ts [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+      const response = await request(app).get('/abc');
+      expect(response.status).toBe(400);
+      expect(response.body.details).toContainEqual({
+        field: 'id',
+        message: 'ID must be a positive integer',
+        value: 'abc'
+      });
+    });
+
+    // MISSING TEST: Should test ID "0" edge case - currently no validation for this
```
Comment (dandeliongold) [inline src/middleware/validation.test.ts:120]: **Missing edge case test**

The comment correctly identifies missing validation for ID "0". This test should be added:

```typescript
it('should reject ID "0"', async () => {
  const response = await request(app).get('/0');
  expect(response.status).toBe(400);
  expect(response.body.details).toContainEqual({
    field: 'id',
    message: 'ID must be a positive integer',
    value: '0'
  });
});
```

**Review by dandeliongold — CHANGES_REQUESTED** [review]
## Code Review Summary

This PR adds comprehensive input validation to the Todo API with good security improvements. However, there are several bugs and missing test coverage that need to be addressed before merging.

### ✅ **Strengths**
- **Excellent validation coverage**: Title length, type checking, HTML tag prevention
- **Structured error responses**: Much better than simple error strings
- **Security improvements**: XSS protection through HTML tag filtering
- **Comprehensive test suite**: Good coverage of validation scenarios
- **Documentation**: Clear validation rules in README

### ⚠️ **Issues Found**

#### **Critical Bugs**
1. **Inconsistent HTML regex** in validation.ts (lines 32 vs 42) - creates security vulnerability
2. **Missing error handling** in POST route - potential server crashes
3. **ID validation allows "0"** - may cause issues depending on ID generation strategy

#### **Missing Test Coverage**
- Description HTML validation test (noted in comment but not implemented)
- ID "0" edge case test (noted in comment but not implemented)

#### **Code Quality**
- Good separation of concerns with middleware
- ValidationException class is defined but not used
- Consistent error response format across all endpoints

### **Breaking Changes**
✅ **Well documented** - Error response format change is clearly noted with migration examples

### **Recommendation**
**REQUEST CHANGES** - The validation logic bugs need to be fixed before merging, particularly the HTML regex inconsistency which creates a security gap.

### **Next Steps**
1. Fix the HTML validation regex inconsistency
2. Add missing test coverage for edge cases
3. Consider adding error handling in route handlers
4. Fix ID validation to properly handle "0" case

Once these issues are resolved, this will be an excellent addition to the codebase!
